### ISSUE TITLE
`unnecessary_null_checks` should have tests with object patterns

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#4183]
Child issue of https://github.com/dart-lang/linter/issues/3624

I think there should be a test case like:

```dart
void f(int? a, int? b) {
  (b!, ) = (a, );
}
```

in which the lint rule would report that null-asserting at `b!` is unnecessary because the variable type is `int?`.
